pages/deployments/oidc/oidc_with_aws.md

```diff
@@ -0,0 +1,187 @@
+
+On this page, as part of the [Creating and managing an OIDC provider (console)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html#manage-oidc-provider-console) process, specify the following values for the:
+
+- **Provider URL**: `https://agent.buildkite.com`
+
+- **Audience**: `sts.amazonaws.com`
+
+## Step 2: Create a new (or update an existing) IAM role to use with your pipelines
+
+Creating new or updating existing IAM roles is conducted through your AWS account.
+
+Learn more about how to do this in the [Creating a role using custom trust policies (console)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-custom.html) page of the AWS documentation.
+
+As part of this process:
+
+1. Choose the **Custom trust policy** role type.
+
+1. Copy the following example trust policy in the following JSON code block and paste it into a code editor:
+
+    ```json
+    {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Principal": {
+                    "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"
+                },
+                "Action": "sts:AssumeRoleWithWebIdentity",
+                "Condition": {
+                    "StringEquals": {
+                        "agent.buildkite.com:aud": "sts.amazonaws.com"
+                    },
+                    "StringLike": {
+                        "agent.buildkite.com:sub": "organization\:ORGANIZATION_SLUG\:pipeline:PIPELINE_SLUG\:ref\:REF\:commit\:BUILD_COMMIT\:step\:STEP_KEY"
+                    },
+                    "IpAddress": {
+                        "aws:SourceIp": [
+                            "AGENT_PUBLIC_IP_ONE",
+                            "AGENT_PUBLIC_IP_TWO"
+                        ]
+                    }
+
+                }
+            }
+        ]
+    }
+    ```
+
+1. Modify the following sections of the pasted code snippet accordingly:
+    * `AWS_ACCOUNT_ID` (in `Principal`) is your AWS account ID.
+    * next item
+
+1. In the **Custom trust policy** section,
+
+1. Specify an appropriate **Role name**, for example, `compute-ssm-oidc-example`.
+
+Set the Principal to be federated via the IAM OIDC provider we just created, and add some conditions against using this IAM role for additional security using the format of the OIDC token subject provided by Buildkite.
+
+This allows us to limit use of this IAM role to named Organizations, Pipelines, and optionally branches, commits and steps. Format is
+ `organization:ORGANIZATION_SLUG:pipeline:PIPELINE_SLUG:ref:REF:commit:BUILD_COMMIT:step:STEP_KEY`
+
+You can use a wildcard `*` to replace any subjects you don’t want to set limits on.
+
+You can also add multiple token subjects as a list if you want to use the same IAM role for multiple pipelines, or if you want a variety of branches, commits and steps enabled.
```

Comment:
So this is what I did in the expanded example at the bottom, but there's basically infinite combinations of things we can do here. 


This is Buildkite organisation `example-org` and their pipeline `example-pipeline`, which can ONLY be exchanged for an IAM role when building on the main branch and tagged releases. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:example-pipeline:ref:refs/heads/main:*",
                         "organization:example-org:pipeline:example-pipeline:ref:refs/tags/*:*"
                     ]
                 }
```

This is Buildkite organisation `example-org` and their pipeline `example-pipeline`, which can be exchanged for an IAM role when building on ANY branch or tag. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:example-pipeline:ref:refs:*",
                     ]
                 }
```

This is Buildkite organisation `example-org` and multiple pipelines pipeline `example-pipeline-one`, `example-pipeline-two` and `example-pipeline-three`, which can each use this role when building on any branch or tag. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:example-pipeline-one:ref:refs:*",
                         "organization:example-org:pipeline:example-pipeline-two:ref:refs:*",
                         "organization:example-org:pipeline:example-pipeline-three:ref:refs:*",
                     ]
                 }
```

This is Buildkite organisation `example-org` and allowing any pipeline in their org, which can use this role when building on any branch or tag. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:*:ref:refs:*",
                     ]
                 }
```

Reply:
This is the write up on why that last example wasn't previous secure. We've mitigated the vulnerability so it shouldn't be possible, but maybe let's not recommend that just in case. https://3.basecamp.com/3453178/buckets/3258372/messages/7155932480

Reply:
Thanks, really useful to know, and I hope I've re-written the instructions to capture this information